src/tools/non-promotion-reasons.md

```diff
@@ -3,60 +3,116 @@ title: Fixing type promotion failures
 description: Solutions for cases where you know more about a field's type than Dart can determine.
 ---
 
-This page has information to help you understand
-why type promotion failures occur,
+[Type promotion][] occurs when flow anlaysis can soundly confirm the value of 
+a [nullable type][] is *not null*, and that its value will not change from that point on.
+Many circumstances can weaken a type's soundness, causing type promotion to fail.
+
+This page lists reasons why type promotion failures occur,
 with tips on how to fix them.
-To learn more, check out [Working with nullable fields][ns-fields],
-a section in [Understanding null safety][].
+To learn more, check out the [Understanding null safety][] page.
 
-[issue #2940]: https://github.com/dart-lang/site-www/issues/2940
-[ns-fields]: /null-safety/understanding-null-safety#working-with-nullable-fields
-[Understanding null safety]: /null-safety/understanding-null-safety
 
+## Unsupported language version for field promotion {#language-version}
+
+**The cause:** 
+You're trying to promote a field, but field promotion is language versioned,
+and your code is set to a language version before 3.2.
+
+If you're already using an SDK version >= Dart 3.2,
+your code might still be explicitly targeted for an earlier [language version][].
+This can happen either because:
+
+* Your [`pubspec.yaml`][] declares an SDK constraint with a
+  lower bound below 3.2, or 
+* You have a `//@dart=version` comment at the top of the file,
+  where `version` is lower than 3.2.
+
+**Example:**
+
+{:.bad}
+```dart
+// @dart=3.1
+
+class C {
+  final int? _i; // CONTEXT 
```

Comment:
I don't see any explanation of what this `// CONTEXT` comment means. Consider either:
- Adding an explanation somewhere that these `// CONTEXT` comments are to let the reader know where the context message will point to
- Or maybe just drop all the `// CONTEXT` comments in this file. After all, the user got to here by clicking on a link that was in a context message; they really don't need us to tell them where the context message pointed to.

I would personally favor dropping all the `// CONTEXT` comments.